source/MaterialXGenShader/ShaderNode.cpp

```diff
@@ -383,38 +414,78 @@ void ShaderNode::setPaths(const Node& node, const NodeDef& nodeDef, bool include
     }
 }
 
-void ShaderNode::setValues(const Node& node, const NodeDef& nodeDef, GenContext& context)
+void ShaderNode::createMetadata(const NodeDef& nodeDef, GenContext& context)
 {
-    // Copy input values from the given node
-    for (const ValueElementPtr& nodeValue : node.getActiveValueElements())
+    ShaderMetadataRegistryPtr registry = context.getUserData<ShaderMetadataRegistry>(ShaderMetadataRegistry::USER_DATA_NAME);
+    if (!(registry && registry->getAllMetadata().size()))
     {
-        ShaderInput* input = getInput(nodeValue->getName());
-        ValueElementPtr nodeDefInput = nodeDef.getActiveValueElement(nodeValue->getName());
-        if (input && nodeDefInput)
-        {
-            const string& valueString = nodeValue->getResolvedValueString();
-            std::pair<const TypeDesc*, ValuePtr> enumResult;
-            if (context.getShaderGenerator().remapEnumeration(*nodeDefInput, valueString, enumResult))
-            {
-                input->setValue(enumResult.second);
-            }
-            else if (!valueString.empty())
-            {
-                input->setValue(nodeValue->getResolvedValue());
-            }
+        // Early out if no metadata is registered.
+        return;
+    }
 
-            InputPtr inputElem = nodeValue->asA<Input>();
-            if (inputElem)
+    // Set metadata on the node according to the nodedef attributes.
+    ShaderMetadataVecPtr nodeMetadataStorage = getMetadata();
+    for (const string& nodedefAttr : nodeDef.getAttributeNames())
+    {
+        const ShaderMetadata* metadataEntry = registry->findMetadata(nodedefAttr);
+        if (metadataEntry)
+        {
+            const string& attrValue = nodeDef.getAttribute(nodedefAttr);
+            if (!attrValue.empty())
             {
-                input->setChannels(inputElem->getChannels());
+                ValuePtr value = Value::createValueFromStrings(attrValue, metadataEntry->type->getName());
+                if (!value)
+                {
+                    value = metadataEntry->value;
+                }
+                if (value)
+                {
+                    if (!nodeMetadataStorage)
+                    {
+                        nodeMetadataStorage = std::make_shared<ShaderMetadataVec>();
+                        setMetadata(nodeMetadataStorage);
+                    }
+                    nodeMetadataStorage->push_back(ShaderMetadata(metadataEntry->name, metadataEntry->type, value));
+                }
             }
         }
     }
 
-    // Set implementation specific values.
-    if (_impl)
+    // Set metadata on inputs according to attributes on the nodedef's inputs
+    for (const ValueElementPtr& nodedefPort : nodeDef.getActiveValueElements())
     {
-        _impl->setValues(node, *this, context);
+        ShaderInput* input = getInput(nodedefPort->getName());
```

Comment:
Is it possible to add meta data to outputs ?

Reply:
I don't think so, the OSL spec say it's for shader parameters and nodes. And for MDL we don't even have outputs everything is returned by function return values.